reporter/reporter.go

```diff
@@ -0,0 +1,250 @@
+package reporter
+
+import (
+	"log"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	metrics "github.com/rcrowley/go-metrics"
+	"github.com/wavefronthq/wavefront-sdk-go/application"
+	"github.com/wavefronthq/wavefront-sdk-go/histogram"
+	wf "github.com/wavefronthq/wavefront-sdk-go/senders"
+)
+
+// WavefrontMetricsReporter report go-metrics to wavefront
+type WavefrontMetricsReporter interface {
+	Start()
+	Stop()
+	Report()
+	ErrorsCount() int64
+}
+
+type reporter struct {
+	mutex        sync.Mutex
+	sender       wf.Sender
+	application  application.Tags
+	source       string
+	prefix       string
+	addSuffix    bool
+	interval     time.Duration
+	ticker       *time.Ticker
+	percentiles  []float64              // Percentiles to export from timers and histograms
+	durationUnit time.Duration          // Time conversion unit for durations
+	metrics      map[string]interface{} // for Wavefron specific metrics tyoes, like Histograms
+	errors       chan error
+	errorsCount  int64
+	stop         chan bool
+	autoStart    bool
+}
+
+// Option allow WavefrontReporter customization
+type Option func(*reporter)
+
+// Source tag for metrics
+func Source(source string) Option {
+	return func(args *reporter) {
+		args.source = source
+	}
+}
+
+// ReportingInterval change the default 1 minute reporting interval.
+func ReportingInterval(interval time.Duration) Option {
+	return func(args *reporter) {
+		args.interval = interval
+	}
+}
+
+// DisableAutoStart prevent the Repotorte to start reporting when is cretaed.
+func DisableAutoStart() Option {
+	return func(args *reporter) {
+		args.autoStart = false
+	}
+}
+
+// Prefix for the metrics name
+func Prefix(prefix string) Option {
+	return func(args *reporter) {
+		args.prefix = strings.TrimSuffix(prefix, ".")
+	}
+}
+
+// AddSuffix add a metric suffix based on the metric type ('.count', '.value')
+func AddSuffix(addSuffix bool) Option {
+	return func(args *reporter) {
+		args.addSuffix = addSuffix
+	}
+}
+
+// New create a WavefrontMetricsReporter
+func New(sender wf.Sender, application application.Tags, setters ...Option) WavefrontMetricsReporter {
+	r := &reporter{
+		sender:       sender,
+		application:  application,
+		source:       "mySource",
+		interval:     time.Second * 5,
+		percentiles:  []float64{0.5, 0.75, 0.95, 0.99, 0.999},
+		durationUnit: time.Nanosecond,
+		metrics:      make(map[string]interface{}),
+		addSuffix:    true,
+		errorsCount:  0,
+		autoStart:    true,
+	}
+
+	for _, setter := range setters {
+		setter(r)
+	}
+
+	r.ticker = time.NewTicker(r.interval)
+	r.stop = make(chan bool, 1)
+	r.errors = make(chan error)
+
+	go func() {
+		for error := range r.errors {
+			if error != nil {
+				r.mutex.Lock()
+				r.errorsCount++
+				log.Printf("(%d) reporter error: %v\n", r.errorsCount, error)
```

Comment:
This could get spammy very quickly if the proxy is unavailable etc. I suggest we do two changes:

1. Add a `debug` Option to the reporter to enable logging individual errors.
2. Add a log statement at the beginning of `Report` that logs the error count if it is greater than zero.